text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,409 @@
+   dependencies in offline environments.
+1. Via the platform. Platforms may choose to offer enhanced functionality to
+   more easily distribute dependencies. In the end, the platform just needs to
+   ensure that the dependency metadata is available to the buildpack in the
+   required location.
+
+## Implementation
+### Tools
+To start the implementation, we’ll create some tooling to work with the new
+metadata format. We’ll create some basic tools for authors to add metadata,
+update metadata, and package metadata into compressed archives and images. For
+buildpack users, we’ll create some tooling to fetch metadata, overlay metadata
+from multiple sources, pre-cache assets from metadata, and feed metadata and
+pre-cached assets into `pack build`.
+
+### Library Enhancement
+The next step will be to add new functionality in `libpak` and `packit` for
+consuming metadata from the defined metadata location and for consuming
+dependency assets from this metadata. This can be done in an additive way so
+that existing functionality continues to be supported. Support for
+`buildpack.toml` dependency metadata can be deprecated and removed on a
+separate timeline.
+
+### Update Buildpacks
+When we have library support, we’ll be able to update the Paketo buildpacks
+that are installing dependencies to use the metadata provided. For example, we
+can update the Bellsoft Liberica or Amazon Corretteo buildpacks to install
+their JRE/JDK dependencies using this new metadata.
+
+We propose a flag of `BP_EXTERNAL_METADATA_ENABLED` which defaults to `false`
+for use as buildpacks are being converted. In the default state, this flag
+tells a buildpack to use the metadata included with buildpack.toml. When set to
+`true`, a buildpack should use the new metadata. This can provide a way for
+users to test the new functionality without impacting existing users.
+
+### Write Documentation
+Documentation will be added at each step. Initial tool documentation will be on
+the repositories for the tools. The `libpak` and `packit` libraries will update
+their docs as support is added. As individual buildpacks are updated, the
+`README.md` for each buildpack will be updated with the newly supported
+options. When an entire language family is complete, the information will be
+merged into the official Paketo documentation.
+
+### Buildpack Migration Process
+Once a language family maintainers is comfortable that all buildpacks have
+migrated to the new metadata format and documentation has been updated, the
+language family may begin the deprecation process. This proposal does not
+dictate how that process should be done, but provides some recommendations. The
+language family team ultimately has discretion on how to proceed.
+
+The recommendation of this proposal is to announce the change in the release
+notes and on Slack, providing links to documentation of the new feature. The
+goal of this migration is that there is no loss of functionality for buildpack
```

Comment:
At the language-family level, I agree this is a non-breaking change and so doesn't have to be a major version bump.

At the individual buildpack level, it is a breaking change though. Sometimes consumers rely on component buildpacks directly outside of the language family, and for these users the removal of a dependency is a breaking change and hence I think the component buildpacks that stop incorporating dependencies should have a major bump.

Reply:
>At the individual buildpack level, it is a breaking change though. Sometimes consumers rely on component buildpacks directly outside of the language family, and for these users the removal of a dependency is a breaking change and hence I think the component buildpacks that stop incorporating dependencies should have a major bump.

...but it doesn't remove dependencies, it removes dependency *metadata*, which is different. IMHO.

Dependencies are part of the public interface/contract with users, but I don't think that where and how we store metadata is part of that contract. So if what we do causes dependencies to be removed, then I 100% agree (I think this is even specified in another RFC) that it triggers a major version update, but I see the actual contents of buildpack.toml as internal to the buildpack and subject to change at our discretion.

I'm not advocating that we just remove it all, but I think the next paragraph strikes a reasonable balance on how teams can deal with that.

Also, this whole section is just "recommendations" and each language family can handle as they see fit.